evidence/reactor_test.go

```
@@ -31,129 +33,176 @@ var (
-		require.False(t, rts.reactor.IsRunning())
+		for _, r := range rts.reactors {
+			if r.IsRunning() {
+				require.NoError(t, r.Stop())
+				require.False(t, r.IsRunning())
+			}
+		}
+
+		leaktest.Check(t)
 	})
 
 	return rts
 }
 
-func createTestSuites(t *testing.T, stateStores []sm.Store, chBuf uint) []*reactorTestSuite {
-	t.Helper()
+func (rts *reactorTestSuite) start(t *testing.T) {
+	rts.network.Start(t)
+	require.Len(t, rts.network.RandomNode().PeerManager.Peers(), rts.numStateStores-1,
+		"network does not have expected number of nodes")
+}
 
-	numSStores := len(stateStores)
-	testSuites := make([]*reactorTestSuite, numSStores)
-	evidenceTime := time.Date(2019, 1, 1, 0, 0, 0, 0, time.UTC)
+func (rts *reactorTestSuite) waitForEvidence(t *testing.T, evList types.EvidenceList, ids ...p2p.NodeID) {
+	t.Helper()
 
-	for i := 0; i < numSStores; i++ {
-		logger := log.TestingLogger().With("validator", i)
-		evidenceDB := dbm.NewMemDB()
-		blockStore := &mocks.BlockStore{}
-		blockStore.On("LoadBlockMeta", mock.AnythingOfType("int64")).Return(
-			&types.BlockMeta{Header: types.Header{Time: evidenceTime}},
+	fn := func(pool *evidence.Pool) {
+		var (
+			localEvList []types.Evidence
+			size        int64
 		)
 
-		pool, err := evidence.NewPool(logger, evidenceDB, stateStores[i], blockStore)
-		require.NoError(t, err)
+		// wait till we have at least the amount of evidence
+		// that we expect. if there's more local evidence then
+		// it doesn't make sense to wait longer and a
+		// different assertion should catch the resulting error
+		loops := 0
+		for len(localEvList) < len(evList) {
+			// each evidence should not be more than 500 bytes
+			localEvList, size = pool.PendingEvidence(int64(len(evList) * 500))
+			if loops == 10 {
+				t.Log("current wait status:", "|",
+					"local", len(localEvList), "|",
+					"waitlist", len(evList), "|",
+					"size", size)
+			}
+			if loops == 10000 && size == 0 {
+				assert.Fail(t, "pool seems disconnected")
```

Let's be consistent with `require` opposed to `assert`. Ideally, we fail early.